### ISSUE TITLE
SOAP exception handling is unhelpful

### ISSUE THREAD

**Issue by candlerb** [zatosource/zato#431]
Example:

~~~
$ curl ....
<?xml version='1.0' encoding='UTF-8'?>
<SOAP-ENV:Envelope
  xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"
  xmlns:xsi="http://www.w3.org/1999/XMLSchema-instance"
  xmlns:xsd="http://www.w3.org/1999/XMLSchema">
   <SOAP-ENV:Body>
     <SOAP-ENV:Fault>
     <faultcode>SOAP-ENV:Client</faultcode>
     <faultstring><![CDATA[cid [K04NMVNN6VVPC5DRMJP5RY0WK7SX], faultstring [Traceback (most recent call last):
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/channel.py", line 198, in dispatch
    payload, worker_store, self.simple_io_config, post_data)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/channel.py", line 351, in handle
    params_priority=channel_item.params_pri)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/service/__init__.py", line 423, in update_handle
    raise e
DetachedInstanceError: Instance <InMessage at 0x7f3e69213a10> is not bound to a Session; attribute refresh operation cannot proceed
]]]></faultstring>
      </SOAP-ENV:Fault>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
~~~

Now, I can see the exception (DetachedInstanceError) but it doesn't show the source line within my handler where this problem arises. The only source paths shown are within zato itself.

Debugging then becomes a tedious matter of repeating the problem and adding `self.logger.info()` calls `:-(`

**Comment by dsuch** [comment]
Hm.. it looks it is a regression from 1.1, let me see if anything can be done about it for 2.0.4.

**Comment by candlerb** [comment]
There is a more general problem here - or possibly it's a facet of the same problem.

Here is an example of a SOAP fault I've just got out of zato:

~~~
$ curl --data-binary @..... -H 'Content-Type: text/xml' http://localhost:11223/....snip....
<?xml version='1.0' encoding='UTF-8'?>
<SOAP-ENV:Envelope
  xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"
  xmlns:xsi="http://www.w3.org/1999/XMLSchema-instance"
  xmlns:xsd="http://www.w3.org/1999/XMLSchema">
   <SOAP-ENV:Body>
     <SOAP-ENV:Fault>
     <faultcode>SOAP-ENV:Client</faultcode>
     <faultstring><![CDATA[cid [K048P9BNE4KYK39MV9V03KDMPAHA], faultstring [Traceback (most recent call last):
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/channel.py", line 198, in dispatch
    payload, worker_store, self.simple_io_config, post_data)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/channel.py", line 351, in handle
    params_priority=channel_item.params_pri)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/service/__init__.py", line 423, in update_handle
    raise e
NameError: global name 'validate_buyer' is not defined
]]]></faultstring>
      </SOAP-ENV:Fault>
  </SOAP-ENV:Body>
~~~

It tells me the actual exception (`NameError: global name 'validate_buyer' is not defined`) but gives no clue as to which line in either the service handler or one of its dependent libraries has the problem.

Now, if I look in server.log, I see *two* backtraces ending with the same NameError: one is the real backtrace which actually locates the source of the problem, and the second is the one from the SOAP message which tells you nothing useful.

~~~
2015-04-24 19:56:32,478 - ESC[1;33mWARNINGESC[0m - 4359:Dummy-153 - zato.server.service:22 - Traceb
ack (most recent call last):
  File "/opt/zato/2.0.3/zato-server/src/zato/server/service/__init__.py", line 402, in update_handl
e
    self._invoke(service, channel)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/service/__init__.py", line 345, in _invoke
    service.handle()
  File "/opt/zato/env/qs-1/server1/work/hot-deploy/current/srv_nd1651.py", line 34, in handle
    remote_ip=self.wsgi_environ.get("REMOTE_ADDR", None))
  File "/opt/zato/2.0.3/zato_extra_paths/nd1651/messaging.py", line 72, in receive_message
    process_message(db, m)
  File "/opt/zato/2.0.3/zato_extra_paths/nd1651/messaging.py", line 113, in process_message
    res = validate_buyer(buyer, seller, **m.meta)
NameError: global name 'validate_buyer' is not defined

2015-04-24 19:56:32,479 - ESC[1;31mERRORESC[0m - 4359:Dummy-153 - zato.server.connection.http_soap.
channel:22 - Caught an exception, cid:[K048P9BNE4KYK39MV9V03KDMPAHA], status_code:[500], _format_ex
c:[Traceback (most recent call last):
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/channel.py", line 198, in
dispatch
    payload, worker_store, self.simple_io_config, post_data)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/connection/http_soap/channel.py", line 351, in
handle
    params_priority=channel_item.params_pri)
  File "/opt/zato/2.0.3/zato-server/src/zato/server/service/__init__.py", line 423, in update_handl
e
    raise e
NameError: global name 'validate_buyer' is not defined
]
~~~

So clearly the information is available somewhere. I would suggest that either:

* The SOAP fault returns the real backtrace. That would be the most useful when developing (although arguably in production would leak information about the structure of the program)

or:

* The SOAP fault suppresses the zato backtrace, which is not adding any useful information

**Comment by dsuch** [comment]
Full tracebacks will now be returned. Done in f457e84e26a739dbc030fcb1cd8249c76296fbe3. Backported to support/2.0. Will be released in 2.0.4.
